nexus/db-queries/src/db/datastore/vpc.rs

```diff
@@ -287,22 +294,60 @@ impl DataStore {
         &self,
         opctx: &OpContext,
         authz_project: &authz::Project,
-        vpc: IncompleteVpc,
+        mut vpc: IncompleteVpc,
     ) -> Result<(authz::Vpc, Vpc), Error> {
-        self.project_create_vpc_raw(
-            opctx,
-            authz_project,
-            InsertVpcQuery::new(vpc),
-        )
-        .await
+        // Attempt to insert a VPC, retrying the query if there are no available
+        // VNIs.
+        const N_ATTEMPTS: usize = 3;
+        let mut attempt = 0;
+        while attempt < N_ATTEMPTS {
+            let original_vni = vpc.vni;
+            match self
+                .project_create_vpc_raw(
+                    opctx,
+                    authz_project,
+                    InsertVpcQuery::new(vpc.clone()),
+                )
+                .await
+            {
+                Ok(Some(vpcs)) => return Ok(vpcs),
+                Err(e) => return Err(e),
+                Ok(None) => {
+                    vpc = vpc.with_new_vni();
+                    debug!(
+                        opctx.log,
+                        "VNI exhaustion detected while inserting VPC, retrying";
+                        "attempt" => attempt,
+                        "vpc_name" => %vpc.identity.name,
+                        "original_vni" => ?original_vni,
+                        "new_vni" => ?vpc.vni,
+                    );
+                    attempt += 1;
+                }
+            }
+        }
+
+        // We've failed to find a VNI within our retry limit. We'll return a 503
+        // in this case.
```

Comment:
This is certainly better than what we had before -- I wonder if there's any point to a "slower, more exhaustive search" once we hit this condition.

Don't get me wrong, I'd be happy to land this change as-is and avoid the immediate issues with memory we're seeing, but it seems like it would be frustrating to keep retrying "project create" to try to find a free VNI, on a system that's very heavily loaded (e.g., a high % of VNIs are allocated).

Either way, happy punting on that until later.

Reply:
I guess it depends on what that slower, more exhaustive search looks like. One very easy option would be to increase the size of the search range, either statically or at each failure. We could certainly pick this to be something in between the entire VNI range (which it used to be) and 0.1% of the range (in this PR). Is that feasible?

What would a more exhaustive search look like? I think avoiding the memory pressure would require restructuring the query, which I'd definitely like to defer.

Reply:
So, in a world with 100 buckets, we're basically doing one of the following:

- Expensive full search: Search all 100 buckets for a free bucket in a single query. This runs out of memory for a larger number of buckets.
- This PR: Pick a random index, scan the next ~5 buckets. If we fail, retry 3 times. This observes a fraction of the total number of buckets, but in a not-super-loaded system, works fine. However, at worst case, this only scans 5% of the total buckets (in the case where "random" returns the same starting index 3 times in a row), and at best case, it scans 15% of the total number of buckets. It's possible this happens when there are free buckets.

I think there are some alternatives here:

1. **Building an efficient allocator**. Perhaps we track the "free VNIs" through a set of ranges, and treat it as a proper range-based allocator. This shares a lot of the problems we'd encounter with a memory allocator, but we'd do the ops in SQL. Rather than scanning, we could simply look up the smallest range in our free list, and pop a VNI out of it.

Pros: It's efficient. It's potentially re-usable for other allocations, like IPs. As long as there is space available, we shouldn't fail.
Cons: It's expensive in engineering time to build. It's complex.

2. **Scanning the entire VNI space in batches**. Rather than "randomly" picking three indices, we "randomly" pick the starting index, from which we scan, and simply iterate over the entire VNI space, one window at a time. For example: we randomly start on bucket 50, scan from 50 -> 55, find nothing free, so we then scan from 55 -> 60, find bucket 56 is free, and return it.

Pros: It's simple. As long as there is space available, we shouldn't fail.
Cons: It's slow on a contended system.

Reply:
I'm not even proposing blocking this PR on these features, but I just want us to be aware of the tradeoff in the following conditions:

| Method | Memory Usage | Cost-to-build | Allocates if space is free? | Efficient under contention |
| -- | -- | -- | -- | -- |
| Full search | Too High | N/A (on `main`) | Yes | Not really? | 
| Random Index + Retry | Low | Cheap | Not always | No, but it fails fast. Though a caller will need to retry! |
| Random Index + Scan everything in batches | Low | Cheap | Yes | No, it would loop for a while | 
| Efficient Allocator | Lowish? | Expensive | Yes | It should be! |

Reply:
Thanks for laying those out. The one I was missing was the batch-based approach, which has a lot of appeal!

There's also another option, I think. The VNIs are a global range -- all VPCs share them. They're also a single number. This is in contrast to the IP addresses, where I intended this query to be used. There, there's a _subnet_ we usually need to search within, and also can be overlap in the subnets between different VPCs.

So what if we put an index on VNIs (I think we already have one), and then restructure the query to have a subquery which takes the next available VNI? I'm not sure if we should avoid allocating sequentially for some reason, but other than that, I don't see a downside. The query is not super complicated, something like

```sql
SELECT IF(
    (SELECT COALESCE(max(vni), 1023) + 1 FROM vpc) > 16777216, -- check if the next VNI is beyond the max
    NULL, -- If so, take NULL which will fail with a `NotNullViolation` error
    (SELECT COALESCE(max(vni), 1023) + 1 FROM vpc) -- or insert the next one
) AS "vni"
```

I'm sure it could be improved, but that would work and handle no entries at all and the table being full.

Reply:
The downside of course is that we're mucking with the Diesel query at the level of raw SQL, which always takes longer than I want it to.

Reply:
Oh, duh. This doesn't wrap at all, so we'll never reallocate VNIs at the bottom of the range once they're freed. Silly me. I like the scan-in-batches approach. It's not very far from this PR (just a bit more bookkeeping around the VNI search range), and should avoid the poor client experience of needing a retry.

Reply:
Ok, I've done the exhaustive search in 3be825e15. Let me know if you have other suggestions, and thanks for the eyes  @smklein!

Reply:
Just to confirm -- are we still picking a random starting offset? (That seems like a minor optimization, but one that might spread out 'where allocations occur').

Reply:
Other than confirming that piece -- looks good to me!

Reply:
Yep starts with a random offset the first time the query is constructed. Then we loop over fixed sized ranges until we successfully done a free VNI.